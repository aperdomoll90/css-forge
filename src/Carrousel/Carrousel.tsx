import React from 'react'
import './styles.css'

export default function Carrousel() {
  return (
    <div className='carrousel-wrapper'>Carrousel</div>
  )
}
