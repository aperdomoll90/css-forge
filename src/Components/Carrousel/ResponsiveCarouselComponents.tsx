import styled from 'styled-components'

const breakpoints = {
  xs: 0,
  md: 768,
  lg: 992,
  xl: 1200,
  xxl: 1536,
}

export const devices = {
  xs: `(min-width: ${breakpoints.xs}px)`,
  md: `(min-width: ${breakpoints.md}px)`,
  lg: `(min-width: ${breakpoints.lg}px)`,
  xl: `(min-width: ${breakpoints.xl}px)`,
  xxl: `(min-width: ${breakpoints.xxl}px)`,
}
// DESIGN  ELEMENTS
export const Wrapper = styled.div`
  height: ${({ height }:any) => (height ? height : '100%')};
  width: ${({ width }:any) => (width ? width : '100%')};
  position: relative;
  margin-top:20vh;
  display: flex;
  align-items: center;
  justify-content: center;
  background: white;

  .inactive {
    color: rgba(80, 78, 78, 0.453);
    font-size: 2.5rem;
    display: block;

    :hover {
    color: rgba(80, 78, 78, 0.453);
  }
  }
`
export const CardContainer = styled.div`
  height: 100%;
  width: ${({ cardContainerWidth }) => (cardContainerWidth ? cardContainerWidth : '100%')};
  position: relative;
  display: flex;
  flex-flow: row nowrap;
  overflow-x: scroll;
  scroll-snap-type: both mandatory;
  scroll-snap-align: center;
  scroll-behavior: smooth;
  cursor: ew-resize;
  background: ${({ color }:any) => color && color};
  flex-direction: ${({ orientation }:any) => (orientation && typeof orientation === 'object' ? orientation.default : orientation)};
  ${({ orientation }:any) =>
    orientation &&
    typeof orientation === 'object' &&
    `
       @media ${devices.md} {
         flex-direction:  ${orientation.md};
       }
       @media ${devices.lg} {
         flex-direction: ${orientation.lg};
       }
       @media ${devices.xl} {
         flex-direction: ${orientation.xl};
       }
       `}
  .show {
    transform: translateY(0);
    opacity: 1;
  }
`
export const CarrouselCard = styled.div`
scroll-snap-align: start;
scroll-snap-stop: always;
transform: translateY(-100px);
opacity: 0;
transition: 150ms;
position: relative;
display: flex;
align-items: center;
justify-content: center;

// TO BE DELETED.... THIS WAS USED FOR BUILDING THE COMPONENT
font-size: 1.5em;
text-align: center;
outline: 1px solid black;


/* CONTROLS SIZE OF THE CARDS USING THE cardsPerView PROP */

${({ cardsPerView }:any) => (cardsPerView && typeof cardsPerView === 'object' ? `flex: 0 0 ${100 / cardsPerView.default}%;` : `flex: 0 0 ${100 / cardsPerView}%;`)}
    
    ${({ cardsPerView }:any) =>
      cardsPerView &&
      typeof cardsPerView === 'object' &&
      `
     @media ${devices.md} {
      flex: 0 0 ${100 / cardsPerView.md}%;
     }
     @media ${devices.lg} {
      flex: 0 0 ${100 / cardsPerView.lg}%;
     }
     @media ${devices.xl} {
      flex: 0 0 ${100 / cardsPerView.xl}%;
     }
     `}
  }
`

// flex: 0 0 ${({ cardsPerView }) => (cardsPerView ? 100 / cardsPerView : 100)}%;

const CarouselButton = styled.button`
  position: absolute;
  z-index: 100;
  min-height: 5rem;
  min-width: 5rem;
  font-size: 3.2rem;
  font-weight: 100;
  background: transparent;
  border: none;
  color: rgba(0, 0, 0, 0.571);
  transition: transform 0.1s ease-in-out;
  cursor: pointer;

  :hover {
    color: yellow;
  }
`
export const LeftCarouselButton = styled(CarouselButton)`
  ${({ orientation }) =>
    orientation && typeof orientation === 'object' ? `${orientation.default === 'row' ? 'left: 0; top:auto; transform: rotate(0deg);' : 'top: 0; transform: rotate(90deg);'}` : `${orientation === 'row' ? 'left: 0; top:auto; transform: rotate(0deg);' : 'top: 0; transform: rotate(90deg);'}`}

  ${({ orientation }) =>
    orientation &&
    typeof orientation === 'object' &&
    `
     @media ${devices.md} {
       ${orientation.md === 'row' ? 'left: 0; top:auto; transform: rotate(0deg);' : 'top: 0; transform: rotate(90deg);'};
     }
     @media ${devices.lg} {
        ${orientation.lg === 'row' ? 'left: 0; top:auto; transform: rotate(0deg);' : 'top: 0; transform: rotate(90deg);'};
     }
     `}
`
export const RightCarouselButton = styled(CarouselButton)`
  ${({ orientation }) =>
    orientation && typeof orientation === 'object'
      ? `${orientation.default === 'row' ? 'right: 0; bottom:auto; transform: rotate(0deg);' : 'bottom: 0; transform: rotate(90deg);'}`
      : `${orientation === 'row' ? 'right: 0; bottom:auto; transform: rotate(0deg);' : 'bottom: 0; transform: rotate(90deg);'}`}

  ${({ orientation }) =>
    orientation &&
    typeof orientation === 'object' &&
    `
   @media ${devices.md} {
     ${orientation.md === 'row' ? 'right: 0; bottom:auto; transform: rotate(0deg);' : 'bottom: 0; transform: rotate(90deg);'};
   }
   @media ${devices.lg} {
      ${orientation.lg === 'row' ? 'right: 0; bottom:auto; transform: rotate(0deg);' : 'bottom: 0; transform: rotate(90deg);'};
   }

   `}
`

export const LeftContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;

`
export const RightContainer = styled.div`
  position: absolute;
  z-index: 100;
  display: flex;
  align-items: center;
  justify-content: center;


`

export const Pagination = styled.nav`
  block-size: calc(10% 100%);
  position: absolute;
  z-index: 100;
  bottom: 1rem;
  left: 1rem;
  right: 1rem;
  display: flex;
  margin: 0 auto;
  inline-size: max-content;
  /* DRY: set here the gap between the dots,
    * we will need it later to set the position
    * of the nav::before pseudoelement
    */
  --gap: 1rem;
  gap: var(--gap, 0.5rem);

  ::before,
  a {
    inline-size: 1rem;
    aspect-ratio: 1;
    border-radius: 50%;
    background: #9bc;
  }
  a {
    inline-size: 1rem;
    aspect-ratio: 1;
    border-radius: 50%;
    background: #9bc;
    text-indent: 100%;
    overflow: hidden;
    white-space: nowrap;
    opacity: 0.33;
  }
  a:hover,
  a:focus {
    background: #eee;
  }

  ::before {
    content: '';
    position: absolute;
    /* z-index: 1; */
    /* display: block;
   cursor: not-allowed; */

    /* here we use the above --gap variable.
    * the --slide variable is set in the
    * keyframes below.
    */
    transform: translateX(calc((100% + var(--gap, 0.5rem)) * calc(var(--slide, 1) - 1)));
    animation: dot 1s steps(1, end) forwards;
    /* shhh... here is the magic */
    animation-timeline: slide;
  }
  /* timeline animation */
  @scroll-timeline slide {
    source: selector(#s);
    orientation: inline;
    time-range: 1s;
  }

  @keyframes dot {
    0% {
      --slide: 1;
    }
    12.5% {
      --slide: 2;
    }
    25% {
      --slide: 3;
    }
    37.5% {
      --slide: 4;
    }
    50%,
    100% {
      --slide: 5;
    }
    62.5%,
    100% {
      --slide: 6;
    }
    75%,
    100% {
      --slide: 7;
    }
    87.5%,
    100% {
      --slide: 8;
    }
  }
`
