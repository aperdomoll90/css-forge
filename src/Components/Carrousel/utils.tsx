import * as React from 'react'


function getPrevElement(list:any) {
    const sibling = list[0].previousElementSibling
  
    if (sibling instanceof HTMLElement) {
      return sibling
    }
  
    return sibling
  }
  
  function getNextElement(list:any) {
    const sibling = list[list.length - 1].nextElementSibling
  
    if (sibling instanceof HTMLElement) {
      return sibling
    }
  
    return null
  }

  export default function usePosition(ref) {
    const [prevElement, setPrevElement] = React.useState(null)
    const [nextElement, setNextElement] = React.useState(null)
  
    React.useEffect(() => {
      const element = ref.current
  
      const update = () => {
        const rect = element.getBoundingClientRect()
  
        const visibleElements:any  = Array.from(element.children).filter(child => {
          const childRect = child.getBoundingClientRect()
  
          return childRect.left >= rect.left && childRect.right <= rect.right
        })
  
        if (visibleElements.length > 0) {
          setPrevElement(getPrevElement(visibleElements))
          setNextElement(getNextElement(visibleElements))
        }
      }
  
      update()
  
      element.addEventListener('scroll', update, { passive: true })
  
      return () => {
        element.removeEventListener('scroll', update, { passive: true })
      }
    }, [ref])
  
    const scrollToElement = React.useCallback(
      element => {
        const currentNode = ref.current
  
        if (!currentNode || !element) return
  
        let newScrollPosition
  
        newScrollPosition =
          element.offsetLeft +
          element.getBoundingClientRect().width / 2 -
          currentNode.getBoundingClientRect().width / 2
  
        currentNode.scroll({
          left: newScrollPosition,
          behavior: 'smooth',
        })
      },
      [ref]
    )
  
    const scrollRight = React.useCallback(
      () => scrollToElement(nextElement),
      [scrollToElement, nextElement]
    )
  
    const scrollLeft = React.useCallback(
      () => scrollToElement(prevElement),
      [scrollToElement, prevElement]
    )
  
    return {
      hasItemsOnLeft: prevElement !== null,
      hasItemsOnRight: nextElement !== null,
      scrollRight,
      scrollLeft,
    }
  }