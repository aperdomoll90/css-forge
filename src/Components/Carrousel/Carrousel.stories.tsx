import { Meta, StoryFn } from '@storybook/react'
import Carrousel from './Carrousel'
import StorybookContainer from '../StorybookContainer'

export default {
  title: 'Carrousel',
  component: Carrousel,
  parameters: {
    jest: ['Carrousel.test.tsx'],
  },
} as Meta

const cardInfo = [{ text: 'Frist card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'Last card' }]


const CarrouselTemplate: StoryFn = (args) => (
  <StorybookContainer>
   <Carrousel {...args}>
      <div>First card</div>
      <div>card</div>
      <div>card</div>
      <div>card</div>
      <div>card</div>
      <div>card</div>
      <div>Last card</div>
    </Carrousel>
  </StorybookContainer>
)

export const CarrouselComponent = CarrouselTemplate.bind({})
CarrouselComponent.args = {
    id: 'carrousel-wrapper',
    testId: 'carrousel-wrapper-testId',
    wrapperClass: 'carrousel-wrapper-className',
    cardsContainerClass: 'card-container-className',
    cardsClass: 'carrousel-card-className',
    //   cardContainerWidth: '100%',
    height: '70%',
    width: '95vw',
    pagination: false,
    navigation: true,
    orientation: { default: 'column', md: 'row', lg: 'row' },
    color: 'blue',
    cardsPerView: { default: 2, md: 2, lg: 2 },
    cardsArray: cardInfo, //in leu off children TO BE DELETED
  }
