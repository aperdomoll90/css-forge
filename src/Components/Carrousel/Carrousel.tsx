import React, { useEffect, useRef } from 'react'

import { CardContainer, CarrouselCard, LeftCarouselButton, RightCarouselButton, Pagination, Wrapper } from './ResponsiveCarouselComponents'
import usePosition from './utils'

export default function Carrousel(props) {
  const leftRef = useRef(null)
  const rightRef = useRef(null)
  const cardContainerRef = useRef(null)
  //   const cardContainer = cardContainerRef?.current

  const { id, testId, cardContainerWidth, height, width, wrapperClass, cardsContainerClass, cardClass, pagination, navigation, color, orientation, cardsPerView, children } = props

  const { hasItemsOnLeft, hasItemsOnRight, scrollRight, scrollLeft } = usePosition(cardContainerRef)
  const classes = ['card']

  cardClass && classes.push(cardClass)

  useEffect(() => {
    const leftButton = leftRef.current
    const rightButton = rightRef.current
    const childrenCards = cardContainerRef?.current?.childNodes
    const cards = Array.prototype.slice.call(childrenCards)
    const firstChild = 0
    const lastChild = cards.length - 1

    const observer = new IntersectionObserver(
      entries => {
        entries.forEach(entry => {
          const intersectingIndex = cards.indexOf(entry.target)
          entry.target.classList.add('show')
          //   this adds the animation instead of the top line
          //   entry.target.classList.toggle('show', entry.isIntersecting)

          navigation && intersectingIndex === firstChild && leftButton?.classList.toggle('inactive', entry.isIntersecting)

          navigation && intersectingIndex === lastChild && rightButton?.classList.toggle('inactive', entry.isIntersecting)
        })
      },
      {
        // threshold monitors how much of the child is present in the observer area
        threshold: 0.5,
        // rootMargin controls when to observe from the edge on the parent container
        // + means will preload and - will wait till is that far inside the parent
        // rootMargin: '3000px',

        // root is to set the parent component always needs to be scrollable
        // root:
      }
    )

    //   INFINITE SCROLLING ADDING EXTRA DIVS
    //   const lastCardObserver = new IntersectionObserver(entries => {
    //   const lastCard = entries[0]
    //   if (!lastCard.isIntersecting) return
    //   Array(10).fill().forEach(loadNewCards)
    //   lastCardObserver.unobserve(lastCard.target)
    //   lastCardObserver.observe(document.querySelector('.card:last-child'))
    // }, {})

    // const cardContainer = document.querySelector('.cardContainer')

    // function loadNewCards() {
    //   const card = document.createElement('div')
    //   card.textContent = 'new card'
    //   card.classList.add('card')
    //   observer.observe(card)
    //   cardContainer.append(card)
    // }

    // lastCardObserver.observe(lastChild)

    cards.forEach(card => {
      observer.observe(card)
    })
  }, [])

  return (
    <Wrapper
      id={id}
      data-testId={testId}
      // ref={responsiveCarouselRef}
      className={wrapperClass}
      height={height}
      width={width}>
      {navigation && (
        <>
          <LeftCarouselButton ref={leftRef} data-testId={'left-button-container'} orientation={orientation} hasItemsOnLeft={hasItemsOnLeft} onClick={scrollLeft} aria-label='Previous slide'>
            &#10094;
          </LeftCarouselButton>

          <RightCarouselButton ref={rightRef} data-testId={'right-button-container'} orientation={orientation}  hasItemsOnRight={hasItemsOnRight} onClick={scrollRight} aria-label='Next slide'>
            &#10095;
          </RightCarouselButton>
        </>
      )}

      <CardContainer ref={cardContainerRef} className={cardsContainerClass} color={color} orientation={orientation} cardContainerWidth={cardContainerWidth}>
        {children &&
          children.map((card, index) => (
            <CarrouselCard key={index} id={`card${index}`} className={classes.join(' ')} cardsPerView={cardsPerView}>
              {card}
            </CarrouselCard>
          ))}
      </CardContainer>

      {pagination && (
        <Pagination>
          {children &&
            children.map((card, index) => (
              // needs to be fixed the <a href> does not work as intended
              <a key={index} href={`#card${index}`} title={`title card ${index}`}>
                {card}
              </a>
            ))}
        </Pagination>
      )}
    </Wrapper>
  )
}
