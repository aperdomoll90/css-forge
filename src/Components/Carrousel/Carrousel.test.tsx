import { render } from '@testing-library/react'
import Carrousel from './Carrousel'

describe('Carrousel component', () => {
  // const { axe, toHaveNoViolations } = require('jest-axe')
  // expect.extend(toHaveNoViolations)
  beforeEach(() => {
    // IntersectionObserver isn't available in test environment
    const mockIntersectionObserver = jest.fn()
    mockIntersectionObserver.mockReturnValue({
      observe: () => null,
      unobserve: () => null,
      disconnect: () => null,
    })
    window.IntersectionObserver = mockIntersectionObserver
  })

  const cardInfo = [{ text: 'Frist card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'card' }, { text: 'Last card' }]

  const args = {
    id: 'carrousel-wrapper',
    testId: 'carrousel-wrapper-testId',
    wrapperClass: 'carrousel-wrapper-className',
    cardsContainerClass: 'card-container-className',
    cardsClass: 'carrousel-card-className',
    //   cardContainerWidth: '100%',
    height: '70%',
    width: '95vw',
    pagination: false,
    navigation: true,
    orientation: { default: 'column', md: 'row', lg: 'row' },
    color: 'blue',
    cardsPerView: { default: 2, md: 2, lg: 2 },
    cardsArray: cardInfo, //in leu off children TO BE DELETED
  }
  const renderComponent = () =>
    render(
      <Carrousel {...args}>
        <div>First card</div>
        <div>card</div>
        <div>card</div>
        <div>card</div>
        <div>card</div>
        <div>card</div>
        <div>Last card</div>
      </Carrousel>
    )
  it('should match snapshots', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
  //   it('should demonstrate ADA compliance on this component', async () => {
  //     const { container } = renderComponent()
  //     const results = await axe(container)
  //     expect(container).toHaveNoViolations()
  //   })
})
