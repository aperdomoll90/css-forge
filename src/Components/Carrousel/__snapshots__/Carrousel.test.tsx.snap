// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carrousel component should match snapshots 1`] = `
<div>
  <div
    class="sc-bgqQcB gZxamG carrousel-wrapper-className"
    data-testid="carrousel-wrapper-testId"
    height="70%"
    id="carrousel-wrapper"
    width="95vw"
  >
    <button
      aria-label="Previous slide"
      class="sc-fFGjHI sc-dicizt kciuDk jCXwCt"
      data-testid="left-button-container"
      orientation="[object Object]"
    >
      ❮
    </button>
    <button
      aria-label="Next slide"
      class="sc-fFGjHI sc-iVCKna kciuDk ieyeCm"
      data-testid="right-button-container"
      orientation="[object Object]"
    >
      ❯
    </button>
    <div
      class="sc-gTRrQi bkHQDV card-container-className"
      color="blue"
      orientation="[object Object]"
    >
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card0"
      >
        <div>
          First card
        </div>
      </div>
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card1"
      >
        <div>
          card
        </div>
      </div>
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card2"
      >
        <div>
          card
        </div>
      </div>
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card3"
      >
        <div>
          card
        </div>
      </div>
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card4"
      >
        <div>
          card
        </div>
      </div>
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card5"
      >
        <div>
          card
        </div>
      </div>
      <div
        class="sc-ewnqHT fLsiAa card"
        id="card6"
      >
        <div>
          Last card
        </div>
      </div>
    </div>
  </div>
</div>
`;
